File: sdks/java/src/dev/ironauth/verify/SampleHarness.java
// SPDX-License-Identifier: MIT OR Apache-2.0
package dev.ironauth.verify;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.interfaces.EdECPublicKey;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * Runs {@link Sample} end to end against a loopback issuer (issue #118).
 *
 * <p>{@link Sample} is the artifact criterion 4 actually asks for: a Java sample that verifies an
 * IronAuth token out of the box. A sample nobody executes is the failure mode worth avoiding
 * here -- it compiles, it reads correctly, and it is wrong. So this stands up a real HTTP server
 * on the loopback interface publishing a real discovery document and a real JWK Set, mints a
 * real token, and drives the sample's whole path: discovery, {@code jwks_uri}, key decode,
 * algorithm allow-list, verification.
 *
 * <p>It uses only {@code com.sun.net.httpserver}, which ships with the JDK, so the no-dependency
 * claim survives its own test. Run as {@code java dev.ironauth.verify.SampleHarness}.
 */
public final class SampleHarness {

    private SampleHarness() {}

    private static final List<String> FAILURES = new ArrayList<>();
    private static int checked;

    /** The algorithms the fake issuer advertises; a test mutates this to include {@code none}. */
    private static volatile String advertisedAlgorithms = "[\"EdDSA\"]";

    /** The issuer name the discovery document claims, which is not always the one being asked for. */
    private static volatile String advertisedIssuer;

    /** When set, discovery points at a key set far larger than the sample will read. */
    private static volatile boolean oversizeJwks;

    /** When set, discovery answers 302 to a document that names the RIGHT issuer and wrong keys. */
    private static volatile boolean redirectDiscovery;

    public static void main(String[] args) throws Exception {
        KeyPair pair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        String x = Fixtures.encodeEd25519((EdECPublicKey) pair.getPublic());
        // A second key, published only behind the redirect, so following it changes the answer.
        String elsewhereX = Fixtures.encodeEd25519(
                (EdECPublicKey) KeyPairGenerator.getInstance("Ed25519").generateKeyPair().getPublic());

        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        String base = "http://127.0.0.1:" + server.getAddress().getPort();
        advertisedIssuer = base;

        server.createContext("/.well-known/openid-configuration", exchange -> {
            if (redirectDiscovery) {
                exchange.getResponseHeaders().add("location", base + "/elsewhere-discovery");
                exchange.sendResponseHeaders(302, -1);
                exchange.close();
                return;
            }
            respond(exchange,
                    "{\"issuer\":\"" + advertisedIssuer + "\",\"jwks_uri\":\"" + base
                            + (oversizeJwks ? "/huge-jwks" : "/jwks") + "\","
                            + "\"id_token_signing_alg_values_supported\":" + advertisedAlgorithms + "}");
        });
        // The redirect target names the CORRECT issuer and points at a DIFFERENT key set. That
        // combination is the point: the issuer check cannot save you here, so only the decision
        // not to follow the redirect does.
        // Two megabytes of valid JSON, over the sample's one-megabyte ceiling.
        server.createContext("/huge-jwks", exchange -> respond(exchange,
                "{\"keys\":[],\"padding\":\"" + "A".repeat(2 << 20) + "\"}"));
        server.createContext("/elsewhere-discovery", exchange -> respond(exchange,
                "{\"issuer\":\"" + base + "\",\"jwks_uri\":\"" + base + "/elsewhere-jwks\","
                        + "\"id_token_signing_alg_values_supported\":[\"EdDSA\"]}"));
        server.createContext("/elsewhere-jwks", exchange -> respond(exchange,
                "{\"keys\":[{\"kty\":\"OKP\",\"crv\":\"Ed25519\",\"x\":\"" + elsewhereX + "\",\"kid\":\"sample-1\"}]}"));
        server.createContext("/jwks", exchange -> respond(exchange,
                "{\"keys\":[{\"kty\":\"OKP\",\"crv\":\"Ed25519\",\"x\":\"" + x + "\",\"kid\":\"sample-1\"}]}"));
        server.start();

        try {
            long now = Instant.now().getEpochSecond();
            String claims = "{\"iss\":\"" + base + "\",\"aud\":\"cli_sample\",\"sub\":\"usr_sample\",\"exp\":"
                    + (now + 3600) + ",\"nbf\":" + (now - 60) + "}";
            String header = "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"sample-1\"}";
            String token = Fixtures.mint(pair, header, claims);

            // THE SAMPLE ACTUALLY RUNS. Discovery, jwks_uri, key decode, verification.
            check("the sample verifies a live Ed25519 token end to end", () -> {
                Map<String, Object> verified = Sample.verify(base, "cli_sample", token);
                return "usr_sample".equals(verified.get("sub")) ? null : "sub was " + verified.get("sub");
            });

            // A tampered token over the same live path, so the harness cannot pass by never
            // reaching the verifier at all.
            String tampered = token.substring(0, token.length() - 4) + "AAAA";
            check("a tampered token is refused over the same path", () -> {
                try {
                    Sample.verify(base, "cli_sample", tampered);
                    return "it verified";
                } catch (IronAuthVerifier.VerifyException refused) {
                    return refused.reason() == IronAuthVerifier.Reason.SIGNATURE_INVALID
                            ? null : "refused as " + refused.reason();
                }
            });

            // An issuer that advertises `none` must not talk the sample into accepting an
            // unsigned token. Note WHERE that is enforced: the refusal below is ALG_NONE, which
            // the verifier raises by name, so this passes even with the sample's own
            // `algorithms.remove("none")` deleted. It is still worth running -- it is the
            // end-to-end statement that a compromised issuer gets nowhere -- but it does not
            // measure that one line, and a mutation run said so.
            advertisedAlgorithms = "[\"none\",\"EdDSA\"]";
            String unsigned = Fixtures.b64("{\"alg\":\"none\",\"typ\":\"JWT\"}".getBytes(StandardCharsets.UTF_8))
                    + "." + Fixtures.b64(claims.getBytes(StandardCharsets.UTF_8)) + ".";
            check("an issuer advertising `none` still cannot get an unsigned token accepted", () -> {
                try {
                    Sample.verify(base, "cli_sample", unsigned);
                    return "it verified";
                } catch (IronAuthVerifier.VerifyException refused) {
                    return refused.reason() == IronAuthVerifier.Reason.ALG_NONE
                            ? null : "refused as " + refused.reason();
                }
            });
            // And the control: with `none` advertised, a GOOD token still verifies. Without this,
            // the assertion above would also pass if advertising `none` broke the sample outright.
            check("with `none` advertised, an honest token still verifies", () -> {
                Map<String, Object> verified = Sample.verify(base, "cli_sample", token);
                return "usr_sample".equals(verified.get("sub")) ? null : "sub was " + verified.get("sub");
            });
            // THE ALLOW-LIST REALLY COMES FROM DISCOVERY. An issuer publishing only RS256 must
            // refuse the honest EdDSA token, on the allow-list and not on the signature. Without
            // this, replacing the discovered list with a hard-coded superset would pass every
            // other check here, and the claim that the issuer decides would be decoration.
            advertisedAlgorithms = "[\"RS256\"]";
            check("an algorithm the issuer does not publish is refused on the allow-list", () -> {
                try {
                    Sample.verify(base, "cli_sample", token);
                    return "it verified";
                } catch (IronAuthVerifier.VerifyException refused) {
                    return refused.reason() == IronAuthVerifier.Reason.ALG_NOT_ALLOWED
                            ? null : "refused as " + refused.reason();
                }
            });
            advertisedAlgorithms = "[\"EdDSA\"]";

            // Discovery that names a different issuer must be refused. Otherwise pointing the
            // sample at any URL yields a document naming an attacker-chosen issuer and a key set
            // to match, and every later comparison passes against that name.
            advertisedIssuer = "https://attacker.example";
            check("discovery naming a different issuer is refused", () -> {
                try {
                    Sample.verify(base, "cli_sample", token);
                    return "it verified";
                } catch (IllegalStateException expected) {
                    return null;
                }
            });
            advertisedIssuer = base;

            // The document size ceiling is REAL, not a length check after the fact: a body is read
            // in bounded chunks and refused the moment it passes the limit. Measured here because
            // a bound nothing ever exceeds is indistinguishable from a bound that does not work.
            oversizeJwks = true;
            check("an oversized key set is refused rather than buffered", () -> {
                try {
                    Sample.verify(base, "cli_sample", token);
                    return "it verified";
                } catch (java.io.IOException expected) {
                    return expected.getMessage().contains("more than") ? null : "failed with " + expected.getMessage();
                }
            });
            oversizeJwks = false;

            // A redirect on discovery is an invitation to fetch someone else's keys, so the
            // client is built with Redirect.NEVER. Following it here would reach a document
            // naming the right issuer and the wrong keys, which every later check would accept.
            // Not following makes the 302 itself the error, and THAT is what this pins.
            redirectDiscovery = true;
            check("a redirect on discovery is not followed", () -> {
                try {
                    Sample.verify(base, "cli_sample", token);
                    return "it verified";
                } catch (java.io.IOException expected) {
                    return expected.getMessage().contains("302") ? null : "failed with " + expected.getMessage();
                }
            });
            redirectDiscovery = false;

            // Plaintext off the loopback interface is refused before any request is made, so this
            // needs no network and cannot flake.
            check("plaintext discovery off loopback is refused", () -> {
                try {
                    Sample.verify("http://issuer.example", "cli_sample", token);
                    return "it verified";
                } catch (IllegalStateException expected) {
                    return null;
                }
            });

            if (checked < 9) {
                FAILURES.add("only " + checked + " checks ran; this harness is its list");
            }
        } finally {
            server.stop(0);
        }

        if (!FAILURES.isEmpty()) {
            System.err.println("FAIL: the Java sample does not hold up end to end");
            FAILURES.forEach(failure -> System.err.println("  - " + failure));
            System.exit(1);
        }
        System.out.println("java sample harness: " + checked + " end-to-end checks against a live issuer OK");
    }

    /** Run one check; the supplier returns null on success or a description of what went wrong. */
    private static void check(String what, ThrowingSupplier check) {
        checked++;
        try {
            String problem = check.get();
            if (problem != null) {
                FAILURES.add(what + " -- " + problem);
            }
        } catch (Exception unexpected) {
            FAILURES.add(what + " -- threw " + unexpected);
        }
    }

    /** A {@link Supplier} that may throw, since every one of these does I/O or crypto. */
    private interface ThrowingSupplier {
        String get() throws Exception;
    }

    private static void respond(HttpExchange exchange, String body) throws java.io.IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().add("content-type", "application/json");
        exchange.sendResponseHeaders(200, bytes.length);
        try (var out = exchange.getResponseBody()) {
            out.write(bytes);
        }
    }
}

File: sdks/dotnet/IronAuth.Verify.Checks/SampleHarness.cs
// SPDX-License-Identifier: MIT OR Apache-2.0
using System.Globalization;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using IronAuth.Verify;
using Org.BouncyCastle.Crypto;

namespace IronAuth.Verify.Checks;

/// <summary>
/// Runs <see cref="Sample"/> end to end against a loopback issuer (issue #118).
/// </summary>
/// <remarks>
/// <see cref="Sample"/> is the artifact criterion 5 actually asks for. A sample nobody executes is
/// the failure mode worth avoiding: it compiles, it reads correctly, and it is wrong. So this
/// stands up a real HTTP listener publishing a real discovery document and a real JWK Set, mints a
/// real token, and drives the sample's whole path: discovery, <c>jwks_uri</c>, key decode,
/// algorithm allow-list, verification.
/// </remarks>
internal static class SampleHarness
{
    private static readonly List<string> Failures = [];
    private static int _checked;

    private static string _advertisedAlgorithms = """["EdDSA"]""";
    private static string _advertisedIssuer = string.Empty;
    private static bool _oversizeJwks;
    private static bool _redirectDiscovery;

    internal static async Task<int> RunAsync()
    {
        AsymmetricCipherKeyPair pair = Fixtures.GenerateEd25519();
        string x = Fixtures.PublicX(pair);
        // A second key, published only behind the redirect, so following it changes the answer.
        string elsewhereX = Fixtures.PublicX(Fixtures.GenerateEd25519());

        int port = FreePort();
        string simpleBase = string.Create(CultureInfo.InvariantCulture, $"http://127.0.0.1:{port}");
        _advertisedIssuer = simpleBase;

        using HttpListener listener = new();
        listener.Prefixes.Add(simpleBase + "/");
        listener.Start();
        using CancellationTokenSource stopping = new();
        Task serving = ServeAsync(listener, simpleBase, x, elsewhereX, stopping.Token);

        // Redirect.NEVER is not the default in .NET either, so the sample builds its own handler.
        using HttpClientHandler handler = new() { AllowAutoRedirect = false };
        using HttpClient http = new(handler) { Timeout = TimeSpan.FromSeconds(5) };

        try
        {
            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            string claims = string.Create(CultureInfo.InvariantCulture,
                $$"""{"iss":"{{simpleBase}}","aud":"cli_sample","sub":"usr_sample","exp":{{now + 3600}},"nbf":{{now - 60}}}""");
            string token = Fixtures.Mint(pair, """{"alg":"EdDSA","typ":"JWT","kid":"sample-1"}""", claims);

            // THE SAMPLE ACTUALLY RUNS. Discovery, jwks_uri, key decode, verification.
            await CheckAsync("the sample verifies a live Ed25519 token end to end", async () =>
            {
                JsonElement verified = await Sample.VerifyAsync(http, simpleBase, "cli_sample", token, now).ConfigureAwait(false);
                return verified.GetProperty("sub").GetString() == "usr_sample" ? null : "sub was wrong";
            }).ConfigureAwait(false);

            // A tampered token over the same live path, so the harness cannot pass by never
            // reaching the verifier at all.
            string tampered = token[..^4] + "AAAA";
            await CheckAsync("a tampered token is refused over the same path", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", tampered, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (VerifyException refused)
                {
                    return refused.Reason == RejectReason.SignatureInvalid ? null : $"refused as {refused.Reason}";
                }
            }).ConfigureAwait(false);

            // An issuer advertising `none` must not talk the sample into accepting an unsigned
            // token. Note WHERE that is enforced: the refusal is AlgNone, which the verifier raises
            // by name, so this passes even with the sample's own `algorithms.Remove("none")`
            // deleted. Still worth running as the end-to-end statement, but it does not measure
            // that one line, and the Java sibling's mutation run said so.
            _advertisedAlgorithms = """["none","EdDSA"]""";
            string unsigned = Base64Url.Encode(Encoding.UTF8.GetBytes("""{"alg":"none","typ":"JWT"}"""))
                + "." + Base64Url.Encode(Encoding.UTF8.GetBytes(claims)) + ".";
            await CheckAsync("an issuer advertising `none` still cannot get an unsigned token accepted", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", unsigned, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (VerifyException refused)
                {
                    return refused.Reason == RejectReason.AlgNone ? null : $"refused as {refused.Reason}";
                }
            }).ConfigureAwait(false);

            // THE ALLOW-LIST REALLY COMES FROM DISCOVERY. An issuer publishing only RS256 must
            // refuse the honest EdDSA token, on the allow-list and not on the signature. Without
            // this, replacing the discovered list with a hard-coded superset would pass every other
            // check here, and the claim that the issuer decides would be decoration.
            _advertisedAlgorithms = """["RS256"]""";
            await CheckAsync("an algorithm the issuer does not publish is refused on the allow-list", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", token, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (VerifyException refused)
                {
                    return refused.Reason == RejectReason.AlgNotAllowed ? null : $"refused as {refused.Reason}";
                }
            }).ConfigureAwait(false);
            _advertisedAlgorithms = """["EdDSA"]""";

            // Discovery naming a different issuer must be refused: otherwise pointing the sample at
            // any URL yields a document naming an attacker-chosen issuer and a key set to match.
            _advertisedIssuer = "https://attacker.example";
            await CheckAsync("discovery naming a different issuer is refused", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", token, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (InvalidOperationException)
                {
                    return null;
                }
            }).ConfigureAwait(false);
            _advertisedIssuer = simpleBase;

            // The document ceiling is REAL, not a length check after the fact: the body is read in
            // bounded chunks and refused the moment it passes the limit. Measured here because a
            // bound nothing ever exceeds is indistinguishable from a bound that does not work.
            _oversizeJwks = true;
            await CheckAsync("an oversized key set is refused rather than buffered", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", token, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (HttpRequestException expected)
                {
                    return expected.Message.Contains("more than", StringComparison.Ordinal) ? null : expected.Message;
                }
            }).ConfigureAwait(false);
            _oversizeJwks = false;

            // A redirect on discovery is an invitation to fetch someone else's keys. The redirect
            // target names the CORRECT issuer and points at a DIFFERENT key set, so the issuer
            // check cannot save us here: only the decision not to follow it can.
            _redirectDiscovery = true;
            await CheckAsync("a redirect on discovery is not followed", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, simpleBase, "cli_sample", token, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (HttpRequestException expected)
                {
                    return expected.Message.Contains("302", StringComparison.Ordinal) ? null : expected.Message;
                }
            }).ConfigureAwait(false);
            _redirectDiscovery = false;

            // Plaintext off the loopback interface is refused before any request, so this needs no
            // network and cannot flake.
            await CheckAsync("plaintext discovery off loopback is refused", async () =>
            {
                try
                {
                    await Sample.VerifyAsync(http, "http://issuer.example", "cli_sample", token, now).ConfigureAwait(false);
                    return "it verified";
                }
                catch (InvalidOperationException)
                {
                    return null;
                }
            }).ConfigureAwait(false);

            if (_checked < 8)
            {
                Failures.Add($"only {_checked} checks ran; this harness is its list");
            }
        }
        finally
        {
            await stopping.CancelAsync().ConfigureAwait(false);
            listener.Stop();
            // The loop returns on its own once the listener stops; awaiting it here means a real
            // fault inside the server still surfaces rather than being swallowed by the finally.
            await serving.ConfigureAwait(false);
        }

        if (Failures.Count > 0)
        {
            Console.Error.WriteLine("FAIL: the .NET sample does not hold up end to end");
            Failures.ForEach(failure => Console.Error.WriteLine("  - " + failure));
            return 1;
        }
        Console.WriteLine($"dotnet sample harness: {_checked} end-to-end checks against a live issuer OK");
        return 0;
    }

    private static async Task ServeAsync(
        HttpListener listener, string simpleBase, string x, string elsewhereX, CancellationToken token)
    {
        while (!token.IsCancellationRequested)
        {
            HttpListenerContext context;
            try
            {
                context = await listener.GetContextAsync().ConfigureAwait(false);
            }
#pragma warning disable CA1031 // shutdown, not a failure: see below
            catch (Exception)
#pragma warning restore CA1031
            {
                // Stopping the listener is HOW this loop ends, and the exception it produces
                // differs by platform (HttpListenerException, ObjectDisposedException, and an
                // InvalidOperationException on some runtimes). Enumerating them would be a list
                // that goes stale on the next runtime; what actually distinguishes shutdown from
                // a real fault is whether we asked it to stop.
                if (token.IsCancellationRequested || !listener.IsListening)
                {
                    return;
                }
                throw;
            }
            string path = context.Request.Url!.AbsolutePath;
            string body;
            switch (path)
            {
                case "/.well-known/openid-configuration" when _redirectDiscovery:
                    context.Response.StatusCode = 302;
                    context.Response.Headers.Add("location", simpleBase + "/elsewhere-discovery");
                    context.Response.Close();
                    continue;
                case "/.well-known/openid-configuration":
                    body = $$"""
                        {"issuer":"{{_advertisedIssuer}}","jwks_uri":"{{simpleBase}}{{(_oversizeJwks ? "/huge-jwks" : "/jwks")}}",
                         "id_token_signing_alg_values_supported":{{_advertisedAlgorithms}}}
                        """;
                    break;
                case "/jwks":
                    body = $$"""{"keys":[{"kty":"OKP","crv":"Ed25519","x":"{{x}}","kid":"sample-1"}]}""";
                    break;
                // Two megabytes of valid JSON, over the sample's one-megabyte ceiling.
                case "/huge-jwks":
                    body = $$"""{"keys":[],"padding":"{{new string('A', 2 << 20)}}"}""";
                    break;
                // The redirect target: the RIGHT issuer, a DIFFERENT key set.
                case "/elsewhere-discovery":
                    body = $$"""
                        {"issuer":"{{simpleBase}}","jwks_uri":"{{simpleBase}}/elsewhere-jwks",
                         "id_token_signing_alg_values_supported":["EdDSA"]}
                        """;
                    break;
                case "/elsewhere-jwks":
                    body = $$"""{"keys":[{"kty":"OKP","crv":"Ed25519","x":"{{elsewhereX}}","kid":"sample-1"}]}""";
                    break;
                default:
                    context.Response.StatusCode = 404;
                    context.Response.Close();
                    continue;
            }
            byte[] bytes = Encoding.UTF8.GetBytes(body);
            context.Response.ContentType = "application/json";
            context.Response.ContentLength64 = bytes.Length;
            await context.Response.OutputStream.WriteAsync(bytes, token).ConfigureAwait(false);
            context.Response.Close();
        }
    }

    /// <summary>
    /// A port nothing is listening on.
    /// </summary>
    /// <remarks>
    /// HttpListener has no "port 0" mode, so a socket is bound to pick one and closed again. There
    /// is a race in principle; on a loopback interface in a test run it has never mattered, and the
    /// alternative is a hard-coded port that collides with whatever else the machine is doing.
    /// </remarks>
    private static int FreePort()
    {
        using TcpListener probe = new(IPAddress.Loopback, 0);
        probe.Start();
        int port = ((IPEndPoint)probe.LocalEndpoint).Port;
        probe.Stop();
        return port;
    }

    private static async Task CheckAsync(string what, Func<Task<string?>> check)
    {
        _checked++;
        try
        {
            string? problem = await check().ConfigureAwait(false);
            if (problem is not null)
            {
                Failures.Add($"{what} -- {problem}");
            }
        }
#pragma warning disable CA1031 // catching everything is the ASSERTION here, not an oversight
        catch (Exception unexpected)
#pragma warning restore CA1031
        {
            Failures.Add($"{what} -- threw {unexpected.GetType().Name}: {unexpected.Message}");
        }
    }
}
